design/hardware-status.md

```diff
@@ -0,0 +1,144 @@
+
+## Motivation
+
+We need to be able to collect and update enough information about hardware
+devices in a server to monitor their status and identify them for
+maintenance and replacement.
+Valuable information are for example the current status, type, serial number,
+physical location, part number.
+We need to be able to refresh the information about the status of the hardware
+devices and, if a device has been replaced, we need to be able to retrieve the
+new information and update the hardware inventory to provide a constant
+up-to-date vision on the status of the devices.
+
+### Goals
+
+1. Decide which information on the devices to retrieve.
+2. Agree on how to retrieve such information.
+3. Provide a way to manually/automatically collect and update the devices
+information.
+
+### Non-Goals
+
+1. Any update on device data should not trigger any reaction from ironic.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- Depending on the device, we need to collect the following information, when
+applicable:
+    * Capacity
+    * Location
+    * Manufacturer
+    * Model
+    * Part Number
+    * Serial Number
+    * Status
+    * SMART details
+    
+- NVMe drives require special tools to get SMART data (e.g. nvme-cli),
+although [smartmontools](https://www.smartmontools.org/wiki/NVMe_Support)
+supports NVMe devices since version 6.5.
+ 
+- [Hdparm](https://en.wikipedia.org/wiki/Hdparm) can provide SMART data for
+most of drives. It doesn't work with NVMe devices.
+
+- [python-hardware](https://github.com/redhat-cip/hardware) can collect SMART
+data, although it requires code changes to detect and provide info on NVMe
+devices, for example relying directly on nvme tools instead of smartmontools.
+
+- [Ironic Python Agent (IPA)](https://github.com/openstack/ironic-python-agent)
+is already able to collect most of this information.
+Using a subset of the IPA code running inside a container in the hosts where we
+want to collect the drives data from can also allow to provide periodic
+updates on all the required info.
+The new containerized service will talk to Ironic Inspector.
+
+- [Ironic Inspector](https://docs.openstack.org/ironic-inspector/latest/)
+needs to be able to publish up-to-date data at any time, including from active
+and available nodes.
+This feature requires changes in the inspector code that currently accept data
+updates only from nodes in manageable state.
+
+### Risks and Mitigations
+
+1. Upstream Ironic community may reject changes in Ironic Inspector code.
+
+## Design Details
+
+### Work Items
+
+- Add option to detect and provide info on NVMe drives to python-hardware.
+- Extract and modify part of the IPA code to be able to run it in a container
+as a new service.
+- Modify Ironic Inspector code to accept data from active nodes.
+- Package (containerize) the new service based on IPA and the necessary tools,
+specifically for NVMe compatibility.
+- Distribute the new service to the hosts.
```

Comment:
I'm still looking for more detail about this. See my earlier comment.

Reply:
We should keep in mind this would be a container with one of the special flags set to operate as a service on each container. I don't remember the way this is done off-hand but it is what is done for things like the existing system utilities like the node-exporter.

Reply:
I think it needs to be part of a DaemonSet. There are several open questions about what owns managing that DaemonSet, though. We have similar questions about how Ironic is set up in the first place, which @sadasu has been working on.

Reply:
As discussed with @dhellmann, and as he mentioned in his comment, there's already an effort to integrate different services on the nodes, and ensure that the pods are deployed, configured and run correctly.
I will add some details about the problem this point tries to solve and a possible approach.